### PR TITLE
fix: ApplicationSet deletes Application status

### LINKED ISSUE #15513: ApplicationSet deletes Application status
**Describe the bug**

We've noticed that starting with v2.8 applications managed by ApplicationSet are periodically loosing status. The `managedFields` indicates that status was updated by `argocd-applictionset-controller`, so very likely it is the culpite.

**To Reproduce**

Unfortunatelly issue is reproducible only with large number of applications and applicationsets.

**Expected behavior**

Appset should not touch application status.

**Screenshots**

**Version**

```shell
v2.8.2
```

### PR DESCRIPTION
**Pull request by alexmt**
Closes https://github.com/argoproj/argo-cd/issues/15513

PR implements 2 applicationset controller changes that fixes https://github.com/argoproj/argo-cd/issues/15513:

* Appset reconciler updates application informer cache after updating the application. I believe stale informer data is a rootcase of the https://github.com/argoproj/argo-cd/issues/15513
* Appset reconciler modifies application using `patch` instead of `update` which guarantees it won't override status won't be touched.

### CODE REVIEW

**Review by jessesuen — APPROVED** [review]
Minor question about updateCache. Not blocking

**Review by gdsoumya — APPROVED** [review]
LGTM

File: applicationset/controllers/applicationset_controller.go [hunk 1]
```diff
@@ -829,15 +849,18 @@ func (r *ApplicationSetReconciler) removeFinalizerOnInvalidDestination(ctx conte
...
 		if len(newFinalizers) != len(app.Finalizers) {
+			updated := app.DeepCopy()
+			updated.Finalizers = newFinalizers
+			if err := r.Client.Patch(ctx, updated, client.MergeFrom(app)); err != nil {
+				return fmt.Errorf("error updating finalizers: %w", err)
+			}
+			if err := r.updateCache(ctx, updated); err != nil {
+				return fmt.Errorf("error updating app cache: %w", err)
+			}
 			app.Finalizers = newFinalizers
```
Comment (jannfis) [inline applicationset/controllers/applicationset_controller.go:860]: This line is basically dead code now, right? Or does the caller reuse `app` with the new finalizers? I would assume that `updated` and `app` are now diverged, at least when it comes to resourceVersion.

Comment (alexmt) [reply applicationset/controllers/applicationset_controller.go]: Unit test insisted finalizers need to be mutated on parameter. I looked at the code and suspect this is not required but scared to change my current behavior. You are right that  `updated` and `app` are now diverged - fixed by using `updated.DeepCopyInto(app)` to keep behavior as is. Thank you!

File: applicationset/controllers/applicationset_controller.go [hunk 2]
```diff
@@ -663,6 +680,9 @@ func (r *ApplicationSetReconciler) createOrUpdateInCluster(ctx context.Context,
 			found.ObjectMeta.Labels = generatedApp.Labels
 			return controllerutil.SetControllerReference(&applicationSet, found, r.Scheme)
 		})
+		if err == nil {
+			err = r.updateCache(ctx, found)
+		}
```
[the remark this answers is not in the document]
Comment (alexmt) [reply applicationset/controllers/applicationset_controller.go]: It should be best effort, agree - updated code.

**Pull request merged**